x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -241,14 +215,26 @@ public DatafeedConfig(StreamInput in) throws IOException {
             }
         }
         if (in.getVersion().before(Version.V_6_6_0)) {
-            this.query = QUERY_TRANSFORMER.toMap(in.readNamedWriteable(QueryBuilder.class));
-            this.aggregations = AGG_TRANSFORMER.toMap(in.readOptionalWriteable(AggregatorFactories.Builder::new));
+            this.queryProvider = QueryProvider.fromParsedQuery(in.readNamedWriteable(QueryBuilder.class));
+            AggregatorFactories.Builder aggs = in.readOptionalWriteable(AggregatorFactories.Builder::new);
+            if (aggs == null) {
+                this.aggProvider = null;
+            } else {
+                this.aggProvider = AggProvider.fromParsedAggs(aggs);
+            }
+        } else if (in.getVersion().before(Version.V_6_6_2)) {
```

Comment:
6.6.2 is already past feature freeze.  This fix is not considered a blocker, so it would be dangerous to assume it will go into 6.6.2.  Unfortunately there's currently no `V_6_6_3` constant.  Therefore I think it's best to change the condition to `in.getVersion().after(Version.V_6_6_2)`, flip the `if` and `else` blocks around and hold off backporting to the 6.6 branch until after 6.6.2 is released.

Same for `writeTo` below.